client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -554,75 +593,133 @@ private void setHeaders(HttpRequest httpRequest, Header[] requestHeaders) {
-                }
+    private NodeTuple<Iterator<Node>> nextNode(NodeSelector nodeSelector) throws IOException {
+        NodeTuple<List<Node>> nodeTuple = this.nodeTuple;
+        List<Node> hosts = selectHosts(nodeTuple, blacklist, lastNodeIndex, System.nanoTime(), nodeSelector);
+        return new NodeTuple<>(hosts.iterator(), nodeTuple.authCache);
+    }
+
+    /**
+     * Select hosts to try. Package private for testing.
+     */
+    static List<Node> selectHosts(NodeTuple<List<Node>> nodeTuple,
+            Map<HttpHost, DeadHostState> blacklist, AtomicInteger lastNodeIndex,
+            long now, NodeSelector nodeSelector) throws IOException {
+        /*
+         * Sort the nodes into living and dead lists.
+         */
+        List<Node> livingNodes = new ArrayList<>(nodeTuple.nodes.size() - blacklist.size());
+        List<DeadNodeAndRevival> deadNodes = new ArrayList<>(blacklist.size());
+        for (Node node : nodeTuple.nodes) {
+            DeadHostState deadness = blacklist.get(node.getHost());
+            if (deadness == null) {
+                livingNodes.add(node);
+                continue;
             }
-            if (filteredHosts.isEmpty()) {
-                //last resort: if there are no good hosts to use, return a single dead one, the one that's closest to being retried
-                List<Map.Entry<HttpHost, DeadHostState>> sortedHosts = new ArrayList<>(blacklist.entrySet());
-                if (sortedHosts.size() > 0) {
-                    Collections.sort(sortedHosts, new Comparator<Map.Entry<HttpHost, DeadHostState>>() {
-                        @Override
-                        public int compare(Map.Entry<HttpHost, DeadHostState> o1, Map.Entry<HttpHost, DeadHostState> o2) {
-                            return o1.getValue().compareTo(o2.getValue());
-                        }
-                    });
-                    HttpHost deadHost = sortedHosts.get(0).getKey();
-                    logger.trace("resurrecting host [" + deadHost + "]");
-                    nextHosts = Collections.singleton(deadHost);
+            long nanosUntilRevival = now - deadness.getDeadUntilNanos();
+            if (nanosUntilRevival > 0) {
+                livingNodes.add(node);
+                continue;
+            }
+            deadNodes.add(new DeadNodeAndRevival(node, nanosUntilRevival));
+        }
+
+        if (false == livingNodes.isEmpty()) {
+            /*
+             * Normal state: there is at least one living node. Rotate the
+             * list so subsequent requests will prefer the nodes in a
+             * different order then run them through the NodeSelector so it
+             * can have its say in which nodes are ok and their ordering. If
+             * the selector is ok with any over the living nodes then use
+             * them for the request.
+             */
+            // TODO this is going to send more requests to nodes right *after* a node that the selector removes
```

Comment:
Very good point! Can we have a test for this? And think about how to fix it next? :)

On one hand it is very problematic to have a different node selector per request (I am even wondering now why we are exposing that, I forgot :)  ), as if the selector is not consistent, we can't be either when round-robining? 

On the other hand, now we mess up even in case the node selector does the same thing over and over, which is bad.

Maybe we should simply take out the ordering aspect here then and perform the rotate after calling the node selector, and make it explicit that the node selector should only select and never rely on ordering? Maybe have a set instead that makes it even more explicit?

Another unrelated problem is that we have an `AtomicInteger` counter... that's going to overflow at some point. Not sure but maybe while we revise this rotate problem we also have a chance to fix the counter limitation.

Reply:
I'd like to revisit this in a followup.